File: device.py
import time
import subprocess
import json
from datetime import datetime

#device parameters
device_type = 'dishwasher'
house = 'kn04'
DEVICE_ID = 'D01'
state = 0 #0 = device OFF, 1 = device ON. It simiulates a sort of readState() function
previous_state = 0
FINISH = 0 # change to 1 when cycle is finished

# This variable contains the chosen panel, based on 50% on its reputation and 50% based on its score
# (panel score is the sum of these 2 values divided by 2)
chosen = {
  "panel_id": "P01",
  "panel_score": 0,
  "start_time": 0,
}
i = 0
# Index of the row corresponding to the current cycle
cycle = 0
# transaction index
start = 1

# Panels list
panels = ["P01", "P02", "P03", "P04", "P05", "SM01"]

# They need to start listening to panels' channels
panelsRoots = {
    "P01": "NCDGXPLVUPDXRJRNRKKBXIARSOAELRVMR9KRDEVPWNUSRKWMVZBDCLWCESUHYRDYFRQZ9ZTOHMCFHMIRD",
    "P02": "WGVLZDBCMFGCEYA9QRNKVGLA9AZRPUAZAS9OCENFQQGIESVEDVEJNKUQWMOPOTMKTGMM9DAHKYJIVRLWH",
    "P03": "YKDCRHXABZN9NBIHEJCAJUFMGCFUMLIDHQKDBJYURPMRAIGCBCBYVDRMPSIQGRYGHDIS9MBLI9ZPUJEZN",
    "P04": "SATIRVGWQTVOCFJJSIVQIKXTPUVWYQNSXAEY9KJGRKJ9KDKBLXEPMEUJTCSRKEEINEHMXHRORDLRRWQPQ",
    "P05": "KCNKHCHJTXOIVAOZBQACKHPYF9EWOCY9QYUAOBHO9DDHOPJTGFTXMSHXRGHVLBPDQNRCOHAPSKLVSGKXZ",
    "SM01": "WSPFIWRDHBYSSLEOANFHNHOPVEUACGTNXPZPHZNWYVXXWMOWKGWGV9EDVUDKMLCQBTSNTRPTPDKJXLOTJ"
}

#panels reputations (panel_id: reputation_value)
panelsReputations = {
       "P01": 0.0,
       "P02": 0.0,
       "P03": 0.0,
       "P04": 0.0,
       "P05": 0.0
}

# Value is in the interval [0, 1], where 0 is the worst case, 1 the best one
def calculatePanelReputation(estimated_timeserie, real_timeserie):
    # Take only samples of real timeserie corresponding with samples of estimated timeserie
    # (to evaluate two functions with same number of element

    # This has the same length of estimated timeserie.
    reduced_real_timeserie = []

    for element in estimated_timeserie:
        for real_element in real_timeserie:
            if (real_element["Time"] == element["Time"]):
                reduced_real_timeserie.append(real_element)
                break

    area_minimum = 0
    area_difference = 0

    for i in range(0, len(estimated_timeserie) - 1):
        # Functions keep order relation in the current and next sample
        if ((estimated_timeserie[i]["Power"] > reduced_real_timeserie[i]["Power"]) and (
                estimated_timeserie[i + 1]["Power"] > reduced_real_timeserie[i + 1]["Power"])):
            area_minimum = area_minimum + (
                        reduced_real_timeserie[i]["Power"] + reduced_real_timeserie[i + 1]["Power"]) / 2
            area_difference = area_difference + abs(
                (estimated_timeserie[i]["Power"] + estimated_timeserie[i + 1]["Power"] -
                 reduced_real_timeserie[i]["Power"] - reduced_real_timeserie[i]["Power"])) / 2

        elif ((estimated_timeserie[i]["Power"] < reduced_real_timeserie[i]["Power"]) and (
                estimated_timeserie[i + 1]["Power"] < reduced_real_timeserie[i + 1]["Power"])):
            area_minimum = area_minimum + (estimated_timeserie[i]["Power"] + estimated_timeserie[i + 1]["Power"]) / 2
            area_difference = area_difference + abs(
                (estimated_timeserie[i]["Power"] + estimated_timeserie[i + 1]["Power"] -
                 reduced_real_timeserie[i]["Power"] - reduced_real_timeserie[i]["Power"])) / 2

        # Functions don't keep order relation in the current and next sample (in example, f1(t1) > f2(t1) but f1(t2) < f2(t2)).
        # In this case, we calculate the intersect between two segments and calculate area of two triangles that they generate
        else:
            # Straight line params (e = estimated, r = real)
            m_e = float(estimated_timeserie[i + 1]["Power"] - estimated_timeserie[i]["Power"])
            q_e = estimated_timeserie[i]["Power"] - m_e

            m_r = float(reduced_real_timeserie[i + 1]["Power"] - reduced_real_timeserie[i]["Power"])
            q_r = reduced_real_timeserie[i]["Power"] - m_r

            if m_e == m_r:
                m_e = m_e + 0.01
            intersect = (q_r - q_e) / (m_e - m_r) - 1

            # Sum of 2 triangles
            area_difference = area_difference + abs(estimated_timeserie[i]["Power"] - reduced_real_timeserie[i]["Power"]) * intersect / 2 + \
                              abs(estimated_timeserie[i + 1]["Power"] - reduced_real_timeserie[i + 1]["Power"]) * (1 - intersect) / 2

            # Difference between the trapeze with one base equals to the smallest i-th sample and the triangle with base equal to the difference between (i + 1)-th samples
            if estimated_timeserie[i]["Power"] < reduced_real_timeserie[i]["Power"]:
                area_minimum = area_minimum + (estimated_timeserie[i]["Power"] + estimated_timeserie[i + 1]["Power"]) / 2 - \
                               abs(estimated_timeserie[i + 1]["Power"] - reduced_real_timeserie[i + 1]["Power"]) * (1 - intersect) / 2
            else:
                area_minimum = area_minimum + (reduced_real_timeserie[i]["Power"] + reduced_real_timeserie[i + 1]["Power"]) / 2 - \
                               abs(estimated_timeserie[i + 1]["Power"] - reduced_real_timeserie[i + 1]["Power"]) * (1 - intersect) / 2
        # print("Area minimum = " + str(area_minimum))
        # print("Area difference = " + str(area_difference))
    return area_difference / (area_difference + area_minimum)

# This function simultate read of a GPIO pin
def readState():
    return 0

# This function is used to retrieve data from a panel; it needs a panel ID
def getData(panel):
    try:
        received_data = subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/panel_receiver.js " +
                                                panelsRoots[panel]], shell=True)
        output = json.loads(received_data)
        print(output["data"])
        panelsRoots[panel] = output["next_root"]
        return output["data"]
    except:
        print("Exception")
        return None

# This function find the panel with best score
def findBestPanel(start, panelList):

    for panel in panelList:

        # To avoid device stuck on specific panel if it doesn't reply, each panel has to send its PROPOSAL in 30 seconds
        timeout = 30

        # Send REQUEST to next panel
        try:
            print("Sending REQUEST transaction to panel " + panel + "...")
            print(start)
            output = subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " +
                                     DEVICE_ID + " " + panel + " " + str(cycle) + " " + str(start) + " " + "REQUEST" + " " + str(0) + " "
                                              + str(0) + " " + device_type + " " + house], shell=True)

            #deviceTimeserie = output["data"]["power"]
            print("REQUEST transaction sent")
            start = start + 1
        except:
            print("Error sending REQUEST transaction")

        # check timeout
        while timeout > 0:
            data = getData(panel)
            if data is not None:

                if data["device_id"] == DEVICE_ID and data["type"] == "PROPOSAL":
                    print("The socre of the panel " + data["panel_id"] + " is " + str(data["score"]))

                    # average between panel score and its reputation
                    if float((data["score"] + panelsReputations[data["panel_id"]]) / 2) > float((chosen["panel_score"] + panelsReputations[data["panel_id"]]) / 2):
                        chosen["panel_id"] = data["panel_id"]
                        chosen["panel_score"] = data["score"]
                        chosen["start_time"] = data["starting_time"]
                        timeout = 0
            else:
                print("No data received")
                timeout = timeout - 1
            time.sleep(2)

    # When there arent't more panels to query, send ACCEPT to best panel (chosen), DENY to other
    for panel in panelList:
        if panel == chosen["panel_id"]:
            try:
                print("Best panel is " + chosen["panel_id"] + ", with a score of " + str(chosen["panel_score"]))
                print("Sending ACCEPT transaction to panel " + panel + "...")
                subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " +
                     DEVICE_ID + " " + chosen["panel_id"] + " " + str(cycle) + " " + str(start) + " " + "ACCEPT" + " " + str(0) + " " + str(0) + " " + device_type + " " + house], shell=True)
                print("ACCEPT transaction sent")
                start = start + 1
            except:
                print("Error sending ACCEPT transaction")
        else:
            try:
                print("Sending DENY transaction to panel " + panel + "...")
                subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " +
                     DEVICE_ID + " " + chosen["panel_id"] + " " + str(cycle) + " " + str(start) + " " + "DENY" + " " + str(0) + " " + str(0) + " " + device_type + " " + house], shell=True)
                print("DENY transaction sent")
                start = start + 1
            except:
                print("Error sending DENY transaction")


# START
# Choose best panel
findBestPanel(start, panels)

while FINISH != 1:

    previous_state = state

    # Get state of device
    state = readState()

    # If device is turned off, so its state changes from 1 to 0
    if(previous_state == 1 and state == 0):

        # Set the end time of the cycle
        date = datetime.now()
        end_time = date.hour * 3600 + date.minute * 60 + date.second

        # Send transaction to smart meter, asking for grid and panel timeseries in the specified interval
        print("Device is turned off, send transaction to smart meter")
        try:
            print("Sending SM_REQUEST transaction to smart meter " + "SM01" + "...")
            print(start)
            subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " +
                                     DEVICE_ID + " " + chosen["panel_id"] + " " + str(cycle) + " " + str(start) + " " + "SM_REQUEST" + " " +
                                     str(chosen["start_time"]) + " " + str(end_time) + " " + device_type + " " + house], shell=True)
            print("SM_REQUEST transaction sent")
            start = start + 1
        except:
            print("Error sending SM_REQUEST transaction")

        # Retrieve data from smart meter
        smData = getData("SM01")
        while (smData is None) or (smData["device_id"] is not DEVICE_ID):
            smData = getData("SM01")

        # update panel reputation - average between new reputation and current reputation
        panelsReputations[smData["panel_id"]] = (calculatePanelReputation(smData["power"]["grid"], smData["power"]["panel"]) + panelsReputations[smData["panel_id"]]) / 2
        FINISH = 1

    # Device is not started yet
    elif(previous_state == readState() == 0):

        try:
            received_data = subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/panel_receiver.js " +
                                                     panelsRoots[chosen["panel_id"]]], shell=True)
            output = json.loads(received_data)
            print(output["data"])
            panelsRoots[chosen["panel_id"]] = output["next_root"]

            if output["data"]["device_id"] == DEVICE_ID:

                # Panel has sent REVOKE and device is off
                if output["data"]["type"] == "REVOKE" and state == 0:
                    # Send REVOKE ACCEPT
                    try:
                        print("Sending REVOKE_ACCEPT transaction to panel " + output["data"]["panel_id"] + "...")
                        subprocess.check_output([
                                                    "node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " +
                                                    DEVICE_ID + " " + chosen["panel_id"] + " " + str(cycle) + " " + str(start) + " " + "REVOKE_ACCEPT" + " " +
                                                    str(0) + " " + str(0) + " " + device_type + " " + house], shell=True)
                        print("REVOKE_ACCEPT transaction sent")
                        start = start + 1
                    except:
                        print("Error sending REQUEST transaction")

                    print("Panel with ID " + chosen["panel_id"] + " has sent REVOKE. Ask to other panels...")
                    #panels.remove(output["data"]["panel_id"])
                    findBestPanel(start, panels)

                # Panel has sent REVOKE and device is on
                elif output["data"]["type"] == "REVOKE" and state == 1:
                    print("I'm working now!")

                    # Send REVOKE DENY
                    try:
                        print("Sending REVOKE_DENY transaction to panel " + output["data"]["panel_id"] + "...")
                        subprocess.check_output(["node /home/giuseppe/Giuseppe/Codice/Tesi\ Smart\ Grid/smart\_energy\_3/smart\_energy/device.js " + DEVICE_ID + " " +
                                                 chosen["panel_id"] + " " + str(cycle) + " " + str(start) + " " + "REVOKE_DENY" + " " + str(0) + " " + str(0) + " " + device_type + " " + house], shell=True)
                        print("REVOKE_DENY transaction sent")
                        start = start + 1
                    except:
                        print("Error sending REQUEST transaction")

                # Panel has sent INIT
                elif output["data"]["type"] == "PANEL_INIT":
                    print("Panel " + output["data"]["panel_id"] + " is active")
                    print(output["data"])

            else:
                print("Ignoring transaction")
        except:
            print("No more data")

    time.sleep(5)

File: device.js
const Mam = require('./lib/mam.client.js');
const IOTA = require('iota.lib.js');
const moment = require('moment');
const iota = new IOTA({ provider: 'http://192.168.1.34:14265' });

//IOTA constants
const SEED = '9VPJ9BNMWREYLEBEPLQDZXSUAFBGXBOORMBNRNJ9YFKKCCZZAONPFW9HJYRIPE9YBFIZTLWPPH9GGDDOU'
const MODE = 'restricted'; // public, private or restricted
const SIDEKEY = 'AOBDOCCDWZWMDPPGMSU9TTAQBMDL99XUNLWIVLRNOQLHJRZJMRPHKGUQJHDGXISUXFVCKGNDZCLEXV9CA'
const SECURITYLEVEL = 2; // 1, 2 or 3
let key;

var start = 0;
var cycle = 0;
var earliest_start = 0;
var latest_start = 0;
var working_time = 0;
var device_id = '';
var panel_id = '';
var type = '';
var start_time = 0;
var end_time = 0;
var device_type = '';
var house = '';

// Check the arguments
const args = process.argv;
if(args.length < 11)
{
    console.log('Missing parameters');
    process.exit();
}
else
{
    device_id = args[2];
    panel_id = args[3];
    cycle = parseInt(args[4]);
    start = parseInt(args[5]);
    type = args[6];
    start_time = parseInt(args[7])
    end_time = parseInt(args[8])
    device_type = args[9]
    house = args[10]
}

// Initialise MAM State
let mamState = Mam.init(iota, SEED, SECURITYLEVEL, start);

// Set channel mode
if(MODE == 'restricted')
{
    const key = iota.utils.toTrytes(SIDEKEY);
    mamState = Mam.changeMode(mamState, MODE, key);
}
else
{
    mamState = Mam.changeMode(mamState, MODE);
}

// This function allow you to create an IOTA transaction passing json data (packet).
// It will return the message root
const publish = async function(packet) {
    // Create MAM Payload
    const trytes = iota.utils.toTrytes(JSON.stringify(packet));
    const message = Mam.create(mamState, trytes);

    // Save new mamState
    mamState = message.state;
    console.log('Root: ', message.root);
    //console.log('Address: ', message.address);

    // Attach the payload.
    await Mam.attach(message.payload, message.address, 3, 9, 'REQUEST');

    return message.root;
}

// This function allow you to publish a transaction created with the previous function
async function createRequestTransaction(device_id, panel_id, cycle, earliest_start, latest_start, working_time, power)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": "REQUEST",
        "power": power,
        "timestamp": dateTime,
        "earliest_start": earliest_start,
        "latest_start": latest_start,
        "working_time": working_time,
        "device_id": device_id,
        "panel_id": panel_id,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

// Finish transaction
async function createFinishTransaction(device_id, panel_id, cycle)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": "FINISH",
        "timestamp": dateTime,
        "device_id": device_id,
        "panel_id": panel_id,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

// Respnse ACK transaction
async function createResponseAckTransaction(device_id, panel_id, cycle, type)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": type,
        "timestamp": dateTime,
        "device_id": device_id,
        "panel_id": panel_id,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

// INIT transaction
async function createInitTransaction(device_id, panel_id, cycle)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": "DEVICE_INIT",
        "timestamp": dateTime,
        "device_id": device_id,
        "panel_id": panel_id,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

// INIT transaction
async function createRevokeTransaction(device_id, panel_id, cycle, type)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": type,
        "timestamp": dateTime,
        "device_id": device_id,
        "panel_id": panel_id,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

// Smart meter request transaction (in this case earliest_start and latest_start are used as start_time and end_time to get panel timeserie
async function createSM_requestTransaction(device_id, panel_id, cycle, start_time, end_time)
{
    const dateTime = moment().utc().format('DD/MM/YYYY hh:mm:ss');

    //send transaction
    var json =
    {
        "type": "SM_REQUEST",
        "timestamp": dateTime,
        "device_id": device_id,
        "panel_id": panel_id,
        "start_time": start_time,
        "end_time": end_time,
        "request_id": cycle
    };
    const root = await publish(json);
    console.log(json);
}

var mysql = require('mysql')
var con = mysql.createConnection({
  host: "services.pep.it",
  user: "giuseppe",
  password: "google7.11",
  database: "smart_grid"
});

con.connect(function(err) {
  if (err) throw err;

  con.query("SELECT * FROM Timeseries WHERE (Cycle_ID = " + cycle + " AND Device_Type = '" + device_type + "' AND House = '" + house + "')", function (err, result, fields) {
    if (err) throw err;
    //console.log(result[0]);
    cycle = result[0]["Cycle_ID"];
    earliest_start = result[0]["Earliest_Start"]
    latest_start = result[0]["Latest_Start"]
    working_time = result[0]["Working_Time"]
    con.query("SELECT * FROM Power WHERE (Cycle_ID = " + cycle + " AND Device_Type = '" + device_type + "' AND House = '" + house + "')", function (err, result, fields) {
          if (err) throw err;
          //console.log(result[0]);

          if(type == "REQUEST") createRequestTransaction(device_id, panel_id, cycle, earliest_start, latest_start, working_time, JSON.stringify(result));
          else if(type == "FINISH") createFinishTransaction(device_id, panel_id, cycle);
          else if(type == "ACCEPT" || type == "DENY") createResponseAckTransaction(device_id, panel_id, cycle, type);
          else if(type == "SM_REQUEST") createSM_requestTransaction(device_id, panel_id, cycle);
          else if(type == "INIT") createInitTransaction(device_id, panel_id, cycle)
          else if(type == "REVOKE_ACCEPT" || type == "REVOKE_DENY") createRevokeTransaction(device_id, panel_id, cycle, type)

          con.end();
        });
  });
});
